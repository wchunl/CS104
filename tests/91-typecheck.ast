ROOT "" (0.0.0)
|   TOK_FUNCTION "" (5.11.1) function
|   |   TOK_TYPE_ID "" (5.11.1) typeid
|   |   |   TOK_VOID "void" (5.11.1) void
|   |   |   TOK_IDENT "foo" (5.11.6) variable
|   |   TOK_PARAM "(" (5.11.9) param
|   TOK_FUNCTION "" (5.12.1) function
|   |   TOK_TYPE_ID "" (5.12.1) typeid
|   |   |   TOK_VOID "void" (5.12.1) void
|   |   |   TOK_IDENT "foo" (5.12.6) variable
|   |   TOK_PARAM "(" (5.12.10) param
|   |   |   TOK_TYPE_ID "" (5.12.11) typeid
|   |   |   |   TOK_INT "int" (5.12.11) int
|   |   |   |   TOK_IDENT "a" (5.12.15) int variable (5.12.11)
|   TOK_CALL "(" (5.18.5)
|   |   TOK_IDENT "f" (5.18.4) variable
|   TOK_CALL "(" (5.19.6)
|   |   TOK_IDENT "f" (5.19.4) variable
|   |   TOK_IDENT "x" (5.19.7) variable
|   |   |   '+' "+" (5.19.11) int vreg
|   |   |   |   TOK_IDENT "y" (5.19.10) variable
|   |   |   |   TOK_INTCON "3" (5.19.12) int const
|   |   |   TOK_IDENT "z" (5.19.15) variable
|   '+' "+" (5.20.8) int vreg
|   |   TOK_IDENT "foo" (5.20.4) variable
|   |   TOK_IDENT "bar" (5.20.10) variable
|   '=' "=" (5.21.6) int vreg
|   |   TOK_IDENT "a" (5.21.4) int variable (5.12.11)
|   |   '=' "=" (5.21.10) int vreg
|   |   |   TOK_IDENT "b" (5.21.8) variable
|   |   |   '=' "=" (5.21.14) int vreg
|   |   |   |   TOK_IDENT "c" (5.21.12) variable
|   |   |   |   TOK_IDENT "d" (5.21.16) variable
|   '=' "=" (5.22.9) int vreg
|   |   TOK_IDENT "test" (5.22.4) variable
|   |   '+' "+" (5.22.21) int vreg
|   |   |   '+' "+" (5.22.15) int vreg
|   |   |   |   TOK_IDENT "abc" (5.22.11) variable
|   |   |   |   TOK_IDENT "def" (5.22.17) variable
|   |   |   TOK_IDENT "ghi" (5.22.23) variable
|   '+' "+" (5.23.18) int vreg
|   |   '+' "+" (5.23.9) int vreg
|   |   |   TOK_IDENT "this" (5.23.4) variable
|   |   |   '*' "*" (5.23.14) int vreg
|   |   |   |   TOK_INTCON "23" (5.23.11) int const
|   |   |   |   TOK_IDENT "a" (5.23.16) int variable (5.12.11)
|   |   TOK_STRINGCON ""hello"" (5.23.20) string const
|   TOK_WHILE "while" (5.24.4)
|   |   '<' "<" (5.24.13)
|   |   |   TOK_IDENT "a" (5.24.11) int variable (5.12.11)
|   |   |   TOK_IDENT "b" (5.24.15) variable
|   |   '=' "=" (5.24.20) int vreg
|   |   |   TOK_IDENT "f" (5.24.18) variable
|   |   |   '+' "+" (5.24.24) int vreg
|   |   |   |   TOK_IDENT "f" (5.24.22) variable
|   |   |   |   TOK_INTCON "1" (5.24.26) int const
|   TOK_RETURN "return" (5.25.4)
|   |   '+' "+" (5.25.13) int vreg
|   |   |   TOK_INTCON "3" (5.25.11) int const
|   |   |   TOK_INTCON "4" (5.25.15) int const
|   '=' "=" (5.26.9) int vreg
|   |   TOK_INDEX "[" (5.26.5) lval vaddr
|   |   |   TOK_IDENT "a" (5.26.4) int variable (5.12.11)
|   |   |   TOK_IDENT "i" (5.26.6) variable
|   |   TOK_INDEX "[" (5.26.12) lval vaddr
|   |   |   TOK_IDENT "b" (5.26.11) variable
|   |   |   TOK_IDENT "j" (5.26.13) variable
|   TOK_RETURNVOID "return" (5.27.4)
|   TOK_WHILE "while" (5.28.4)
|   |   TOK_IDENT "TRUE" (5.28.11) variable
|   |   TOK_BLOCK "{" (5.28.17)
|   |   |   '=' "=" (5.28.20) int vreg
|   |   |   |   TOK_IDENT "a" (5.28.18) int variable (5.12.11)
|   |   |   |   TOK_INTCON "3" (5.28.22) int const
|   |   |   '=' "=" (5.28.27) int vreg
|   |   |   |   TOK_IDENT "b" (5.28.25) variable
|   |   |   |   TOK_INTCON "4" (5.28.29) int const
|   TOK_IF "if" (5.29.4)  f
|   |   TOK_EQ "==" (5.29.10) vreg
|   |   |   TOK_IDENT "a" (5.29.8) int variable (5.12.11)
|   |   |   TOK_IDENT "b" (5.29.13) variable
|   |   TOK_CALL "(" (5.29.18)
|   |   |   TOK_IDENT "f" (5.29.16) variable
|   |   |   TOK_IDENT "x" (5.29.19) variable
|   TOK_IF "if" (5.30.4)
|   |   TOK_NE "!=" (5.30.10)
|   |   |   TOK_IDENT "a" (5.30.8) int variable (5.12.11)
|   |   |   TOK_IDENT "b" (5.30.13) variable
|   |   '=' "=" (5.30.18) int vreg
|   |   |   TOK_IDENT "y" (5.30.16) variable
|   |   |   TOK_INTCON "3" (5.30.20) int const
|   |   TOK_IFELSE "(" (5.30.30)
|   |   |   TOK_IDENT "f" (5.30.28) variable
|   |   |   TOK_IDENT "y" (5.30.31) variable
|   |   |   |   TOK_IDENT "z" (5.30.34) variable
