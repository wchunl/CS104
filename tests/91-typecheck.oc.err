8X:9.4: syntax error, unexpected '[', expecting TOK_IDENT
8X:10.11: syntax error, unexpected ']'
8X:13.14: syntax error, unexpected '[', expecting TOK_IDENT
8X:16.1: syntax error, unexpected TOK_INT, expecting ';'
8X:31.1: syntax error, unexpected '}'
