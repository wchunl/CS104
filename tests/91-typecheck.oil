foo:      .function
          .end
foo:      .function
          .param int a
          .end
