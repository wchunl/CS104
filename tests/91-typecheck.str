string_set[   2]:    2594088158685378223 0x2165e28->"10"
                   12597418120160561823 0x2161318->"int"
                    3663726644998027833 0x2167398->"abc"
string_set[   4]:    2428694746313403453 0x2167688->"this"
string_set[   5]:    6519678867730460866 0x2167998->""hello""
                    1370730858159036685 0x2166718->"+"
string_set[   7]:    9631199822919835226 0x2165ec8->"foo"
                   10959529184379665549 0x2166ea8->"c"
string_set[   8]:   13497578857207247412 0x2165d18->"string"
string_set[   9]:   10319632053171904911 0x2167ae8->"while"
string_set[  11]:    4993892634952068459 0x2165aa8->"a"
string_set[  12]:   16215888864653804456 0x21681d8->"4"
string_set[  13]:    6253375586064260614 0x2168388->"i"
string_set[  14]:   15816506932648033945 0x2165f18->"("
                    1396094175042499165 0x2167c78->"<"
string_set[  15]:   12868445110721718657 0x2165f68->")"
string_set[  17]:    6142509188972423790 0x2161738->""
string_set[  18]:   14494284460613645429 0x2167178->"d"
string_set[  20]:    4797595651836319732 0x2168628->"TRUE"
string_set[  21]:    9317282556992397945 0x2168bf8->"if"
string_set[  22]:   10159970873491820195 0x2167d88->"1"
string_set[  26]:   11597697714117577063 0x2165ac8->"="
string_set[  28]:    4587476179349545048 0x2165cc8->"new"
string_set[  29]:   11911796782241914524 0x2165e78->"void"
string_set[  31]:   15160494916537613830 0x21663b8->"y"
string_set[  35]:   11868120863286193964 0x2165d68->"main"
string_set[  37]:    9381151195890144283 0x2165b18->"null"
string_set[  38]:   11697390762615875584 0x21674a8->"def"
                   11367455634845683663 0x2167558->"ghi"
string_set[  40]:   13945322250941355849 0x21667b8->"struct"
string_set[  41]:   12051031859636400188 0x2165a58->"]"
                   13257391922330118892 0x2165bb8->"reference"
string_set[  42]:   13686721134187162920 0x2166768->"}"
string_set[  50]:      21606623943142546 0x2166678->"{"
string_set[  51]:   13272411544345499535 0x21666c8->"x"
string_set[  52]:    5344662657899890615 0x2165b68->";"
string_set[  54]:   10838281452030117757 0x2166628->"b"
string_set[  56]:   11474628671133349555 0x2166c88->"bar"
                    8968846175329310707 0x2166958->"f"
string_set[  60]:    1873634190020599143 0x2169258->"else"
string_set[  61]:   14168603451105147218 0x21677d8->"23"
string_set[  67]:    4706206694200543916 0x2166018->","
string_set[  75]:    8248777770799913213 0x2166bd8->"3"
string_set[  77]:   17970277606898470454 0x2166cc8->"z"
string_set[  80]:    2617056156435215158 0x2168f08->"!="
string_set[  83]:   13581035385482130682 0x2161758->"["
string_set[  87]:   14635384904416252381 0x2168038->"return"
string_set[  89]:    6637313742931709005 0x2167888->"*"
string_set[  94]:   15118982290295364091 0x2167228->"test"
string_set[  95]:   14736352883882276958 0x2168df8->"=="
string_set[  96]:   10485857595211860659 0x2168438->"j"
load_factor = 0.526
bucket_count = 97
max_bucket_size = 3
