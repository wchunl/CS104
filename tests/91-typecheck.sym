foo (5.11.1) {0} void function

foo (5.12.1) {1} void function
   a (5.12.11) {1} int variable param lval 0
