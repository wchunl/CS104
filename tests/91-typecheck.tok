#  1 "91-typecheck.oc"
#  1 "<built-in>"
#  1 "<command-line>"
#  1 "91-typecheck.oc"
#  9 "91-typecheck.oc"
   5 9.00001 262 TOK_INT       int
   5 9.00004  91 '['           [
   5 9.00005  93 ']'           ]
   5 9.00007 282 TOK_IDENT     a
   5 9.00009  61 '='           =
   5 9.00011 282 TOK_IDENT     null
   5 9.00015  59 ';'           ;
   5 10.00001 282 TOK_IDENT     reference
   5 10.00010  91 '['           [
   5 10.00011  93 ']'           ]
   5 10.00013 282 TOK_IDENT     a
   5 10.00015  61 '='           =
   5 10.00017 282 TOK_IDENT     new
   5 10.00021 263 TOK_STRING    string
   5 10.00027  91 '['           [
   5 10.00028 283 TOK_INTCON    10
   5 10.00030  93 ']'           ]
   5 10.00031  59 ';'           ;
   5 11.00001 261 TOK_VOID      void
   5 11.00006 282 TOK_IDENT     foo
   5 11.00009  40 '('           (
   5 11.00010  41 ')'           )
   5 11.00011  59 ';'           ;
   5 12.00001 261 TOK_VOID      void
   5 12.00006 282 TOK_IDENT     foo
   5 12.00010  40 '('           (
   5 12.00011 262 TOK_INT       int
   5 12.00015 282 TOK_IDENT     a
   5 12.00016  41 ')'           )
   5 12.00017  59 ';'           ;
   5 13.00001 261 TOK_VOID      void
   5 13.00006 282 TOK_IDENT     foo
   5 13.00010  40 '('           (
   5 13.00011 262 TOK_INT       int
   5 13.00014  91 '['           [
   5 13.00015  93 ']'           ]
   5 13.00017 282 TOK_IDENT     a
   5 13.00018  44 ','           ,
   5 13.00020 262 TOK_INT       int
   5 13.00023  91 '['           [
   5 13.00024  93 ']'           ]
   5 13.00026 282 TOK_IDENT     b
   5 13.00027  41 ')'           )
   5 13.00029 123 '{'           {
   5 13.00030 262 TOK_INT       int
   5 13.00034 282 TOK_IDENT     x
   5 13.00036  61 '='           =
   5 13.00038 282 TOK_IDENT     a
   5 13.00040  43 '+'           +
   5 13.00042 282 TOK_IDENT     b
   5 13.00043  59 ';'           ;
   5 13.00044 125 '}'           }
   5 14.00001 269 TOK_STRUCT    struct
   5 14.00008 282 TOK_IDENT     foo
   5 14.00012 123 '{'           {
   5 14.00014 262 TOK_INT       int
   5 14.00018 282 TOK_IDENT     a
   5 14.00019  59 ';'           ;
   5 14.00021 262 TOK_INT       int
   5 14.00025 282 TOK_IDENT     b
   5 14.00026  59 ';'           ;
   5 14.00028 125 '}'           }
   5 16.00001 262 TOK_INT       int
   5 16.00005 282 TOK_IDENT     main
   5 16.00009  40 '('           (
   5 16.00010  41 ')'           )
   5 16.00012 123 '{'           {
   5 17.00004 282 TOK_IDENT     a
   5 17.00006  43 '+'           +
   5 17.00008 282 TOK_IDENT     b
   5 17.00009  59 ';'           ;
   5 18.00004 282 TOK_IDENT     f
   5 18.00005  40 '('           (
   5 18.00006  41 ')'           )
   5 18.00007  59 ';'           ;
   5 19.00004 282 TOK_IDENT     f
   5 19.00006  40 '('           (
   5 19.00007 282 TOK_IDENT     x
   5 19.00008  44 ','           ,
   5 19.00010 282 TOK_IDENT     y
   5 19.00011  43 '+'           +
   5 19.00012 283 TOK_INTCON    3
   5 19.00013  44 ','           ,
   5 19.00015 282 TOK_IDENT     z
   5 19.00016  41 ')'           )
   5 19.00017  59 ';'           ;
   5 20.00004 282 TOK_IDENT     foo
   5 20.00008  43 '+'           +
   5 20.00010 282 TOK_IDENT     bar
   5 20.00013  59 ';'           ;
   5 21.00004 282 TOK_IDENT     a
   5 21.00006  61 '='           =
   5 21.00008 282 TOK_IDENT     b
   5 21.00010  61 '='           =
   5 21.00012 282 TOK_IDENT     c
   5 21.00014  61 '='           =
   5 21.00016 282 TOK_IDENT     d
   5 21.00017  59 ';'           ;
   5 22.00004 282 TOK_IDENT     test
   5 22.00009  61 '='           =
   5 22.00011 282 TOK_IDENT     abc
   5 22.00015  43 '+'           +
   5 22.00017 282 TOK_IDENT     def
   5 22.00021  43 '+'           +
   5 22.00023 282 TOK_IDENT     ghi
   5 22.00026  59 ';'           ;
   5 23.00004 282 TOK_IDENT     this
   5 23.00009  43 '+'           +
   5 23.00011 283 TOK_INTCON    23
   5 23.00014  42 '*'           *
   5 23.00016 282 TOK_IDENT     a
   5 23.00018  43 '+'           +
   5 23.00020 285 TOK_STRINGCON "hello"
   5 23.00027  59 ';'           ;
   5 24.00004 267 TOK_WHILE     while
   5 24.00010  40 '('           (
   5 24.00011 282 TOK_IDENT     a
   5 24.00013  60 '<'           <
   5 24.00015 282 TOK_IDENT     b
   5 24.00016  41 ')'           )
   5 24.00018 282 TOK_IDENT     f
   5 24.00020  61 '='           =
   5 24.00022 282 TOK_IDENT     f
   5 24.00024  43 '+'           +
   5 24.00026 283 TOK_INTCON    1
   5 24.00027  59 ';'           ;
   5 25.00004 268 TOK_RETURN    return
   5 25.00011 283 TOK_INTCON    3
   5 25.00013  43 '+'           +
   5 25.00015 283 TOK_INTCON    4
   5 25.00016  59 ';'           ;
   5 26.00004 282 TOK_IDENT     a
   5 26.00005  91 '['           [
   5 26.00006 282 TOK_IDENT     i
   5 26.00007  93 ']'           ]
   5 26.00009  61 '='           =
   5 26.00011 282 TOK_IDENT     b
   5 26.00012  91 '['           [
   5 26.00013 282 TOK_IDENT     j
   5 26.00014  93 ']'           ]
   5 26.00015  59 ';'           ;
   5 27.00004 268 TOK_RETURN    return
   5 27.00010  59 ';'           ;
   5 28.00004 267 TOK_WHILE     while
   5 28.00010  40 '('           (
   5 28.00011 282 TOK_IDENT     TRUE
   5 28.00015  41 ')'           )
   5 28.00017 123 '{'           {
   5 28.00018 282 TOK_IDENT     a
   5 28.00020  61 '='           =
   5 28.00022 283 TOK_INTCON    3
   5 28.00023  59 ';'           ;
   5 28.00025 282 TOK_IDENT     b
   5 28.00027  61 '='           =
   5 28.00029 283 TOK_INTCON    4
   5 28.00030  59 ';'           ;
   5 28.00032 125 '}'           }
   5 29.00004 265 TOK_IF        if
   5 29.00007  40 '('           (
   5 29.00008 282 TOK_IDENT     a
   5 29.00010 275 TOK_EQ        ==
   5 29.00013 282 TOK_IDENT     b
   5 29.00014  41 ')'           )
   5 29.00016 282 TOK_IDENT     f
   5 29.00018  40 '('           (
   5 29.00019 282 TOK_IDENT     x
   5 29.00020  41 ')'           )
   5 29.00021  59 ';'           ;
   5 30.00004 265 TOK_IF        if
   5 30.00007  40 '('           (
   5 30.00008 282 TOK_IDENT     a
   5 30.00010 276 TOK_NE        !=
   5 30.00013 282 TOK_IDENT     b
   5 30.00014  41 ')'           )
   5 30.00016 282 TOK_IDENT     y
   5 30.00018  61 '='           =
   5 30.00020 283 TOK_INTCON    3
   5 30.00021  59 ';'           ;
   5 30.00023 266 TOK_ELSE      else
   5 30.00028 282 TOK_IDENT     f
   5 30.00030  40 '('           (
   5 30.00031 282 TOK_IDENT     y
   5 30.00032  44 ','           ,
   5 30.00034 282 TOK_IDENT     z
   5 30.00035  41 ')'           )
   5 30.00036  59 ';'           ;
   5 31.00001 125 '}'           }
