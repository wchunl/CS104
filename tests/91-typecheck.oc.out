; 0.  1: # 1 "91-typecheck.oc"; 1.  0: 
; 1.  1: # 1 "<built-in>"
; 2.  1: # 1 "<command-line>"
; 3.  1: # 1 "91-typecheck.oc"
; 4.  1: # 9 "91-typecheck.oc"
; 5.  9: int[] a = null;
; 5. 10: reference[] a = new string[10];
; 5. 11: void foo();
; 5. 12: void foo (int a);
; 5. 13: void foo (int[] a, int[] b) {int x = a + b;}
; 5. 14: struct foo { int a; int b; }
; 5. 15: 
; 5. 16: int main() {
; 5. 17:    a + b;
; 5. 18:    f();
; 5. 19:    f (x, y+3, z);
; 5. 20:    foo + bar;
; 5. 21:    a = b = c = d;
; 5. 22:    test = abc + def + ghi;
; 5. 23:    this + 23 * a + "hello";
; 5. 24:    while (a < b) f = f + 1;
; 5. 25:    return 3 + 4;
; 5. 26:    a[i] = b[j];
; 5. 27:    return;
; 5. 28:    while (TRUE) {a = 3; b = 4; }
; 5. 29:    if (a == b) f (x);
; 5. 30:    if (a != b) y = 3; else f (y, z);
; 5. 31: }
We don't need TOK_IDENTWe don't need TOK_IDENTWe don't need TOK_IDENTWe don't need TOK_IDENTWe don't need TOK_IDENTWe don't need TOK_IDENTWe don't need TOK_IDENTWe don't need TOK_IDENTWe don't need TOK_IDENTWe don't need TOK_IDENTWe don't need TOK_IDENTWe don't need TOK_IDENTWe don't need TOK_IDENTWe don't need TOK_IDENTWe don't need TOK_IDENTWe don't need TOK_IDENTWe don't need TOK_IDENTWe don't need TOK_IDENTWe don't need TOK_IDENTWe don't need TOK_IDENTWe don't need TOK_IDENTWe don't need TOK_IDENTWe don't need TOK_IDENTWe don't need TOK_IDENTWe don't need TOK_IDENTWe don't need TOK_IDENTWe don't need TOK_IDENTWe don't need TOK_IDENTWe don't need TOK_IDENTWe don't need TOK_IDENTWe don't need TOK_IDENTWe don't need TOK_IDENTWe don't need TOK_IDENTWe don't need TOK_IDENTWe don't need TOK_IDENT